### PR TITLE
mixer: dont stop mixer on pause

### LINKED ISSUE #3522: [BUG] IPC error during pause/start test with multiple mixer sources
_Initially filed as a Linux bug since the result is an IPC error, now transferred to SOF._

I've seen this error multiple times with sof-test, this is not a random error but a mixer issue from the logs.

@lgirdwood @lyakh @singalsu FYI

Log files: [multiple-pause-resume.tar.gz](https://github.com/thesofproject/linux/files/5374776/multiple-pause-resume.tar.gz)

excerpts:
dmesg: 
````
kernel: [  531.157632] sof-audio-acpi 808622A8:00: ipc tx: 0x60040000: GLB_STREAM_MSG: TRIG_START
kernel: [  531.157787] sof-audio-acpi 808622A8:00: ipc tx succeeded: 0x60040000: GLB_STREAM_MSG: TRIG_START
kernel: [  531.317334] sof-audio-acpi 808622A8:00: pcm: trigger stream 0 dir 0 cmd 3
kernel: [  531.317358] sof-audio-acpi 808622A8:00: ipc tx: 0x60060000: GLB_STREAM_MSG: TRIG_PAUSE
kernel: [  531.317486] sof-audio-acpi 808622A8:00: ipc tx succeeded: 0x60060000: GLB_STREAM_MSG: TRIG_PAUSE
kernel: [  531.417219] sof-audio-acpi 808622A8:00: pcm: trigger stream 1 dir 0 cmd 1
kernel: [  531.417276] sof-audio-acpi 808622A8:00: ipc tx: 0x60040000: GLB_STREAM_MSG: TRIG_START
kernel: [  531.417408] sof-audio-acpi 808622A8:00: error: ipc error for 0x60040000 size 12
kernel: [  531.417435] sof-audio-acpi 808622A8:00: ASoC: error at snd_soc_pcm_component_trigger on 808622A8:00: -22
kernel: [  531.417450]  PCM Deep Buffer: ASoC: trigger FE cmd: 1 failed: -22
kernel: [  531.417472] sof-audio-acpi 808622A8:00: pcm: trigger stream 1 dir 0 cmd 0
````

etrace:
````
        TIMESTAMP              DELTA C# COMPONENT          LOCATION                      CONTENT
[  8250820.781250] (  8250821.000000) c0 dma-trace             src/trace/dma-trace.c:327  ERROR FW ABI 0x3012000 DBG ABI 0x5002001 tag v1.5-rc1-778-ge5d17f3c10f3 src hash 0xa3b71c81 (ldc hash 0xa3b71c81\
)
[531415530.364583] (523164704.000000) c0 mixer        1.1      src/audio/component.c:190  ERROR comp_set_state(): wrong state = 4, COMP_TRIGGER_START
[531415540.364583] (       10.000000) c0 pipe         4.25      src/audio/pipeline.c:884  ERROR pipeline_trigger(): ret = -22, host->comp.id = 21, cmd = 1
[531415563.750000] (       23.385416) c0 ipc                       src/ipc/handler.c:443  ERROR ipc: comp 21 trigger 0x40000 failed -22
```

### PR DESCRIPTION
**Pull request by abonislawski**
audio_stream: add function writing zeros to the buffer without reference to a source buffer
mixer: dont stop mixer on pause and write zeros if all sources are inactive

Fixes #3522

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
Will this also when one stream is stopped ? or are we doing something different in this case ?
Btw, is there a similar update for demux ?

**Comment by abonislawski** [comment]
> Will this also when one stream is stopped ? or are we doing something different in this case ?
> Btw, is there a similar update for demux ?

If only one stream is stopped mixer already can handle this, we will only not stop mixer ppl
No info about demux

File: src/audio/mixer.c [hunk 1]
```diff
@@ -245,13 +245,18 @@ static int mixer_trigger(struct comp_dev *dev, int cmd)
...
 	/* don't stop mixer if at least one source is paused */
 	} else if (mixer_source_status_count(dev, COMP_STATE_PAUSED) &&
 		   cmd == COMP_TRIGGER_STOP) {
-		dev->state = COMP_STATE_PAUSED;
+		dev->state = COMP_STATE_ACTIVE;
+		ret = PPL_STATUS_PATH_STOP;
+	}
+	/* don't stop mixer on pause */
+	if (cmd == COMP_TRIGGER_PAUSE) {
+		dev->state = COMP_STATE_ACTIVE;
```
Comment (plbossart) [inline src/audio/mixer.c:259]: you may want to reflow the logic as

````c
if ((cmd == COMP_TRIGGER_STOP) && 
            (mixer_source_status_count(dev, COMP_STATE_ACTIVE) ||
            mixer_source_status_count(dev, COMP_STATE_PAUSED)) ||
    (cmd == COMP_TRIGGER_PAUSE)) {
                dev->state = COMP_STATE_ACTIVE;
		ret = PPL_STATUS_PATH_STOP;
}
````

or use a switch case with a fallback.




￼

**Review by plbossart — CHANGES_REQUESTED** [review]
the logic can be improved a bit?

File: src/audio/mixer.c (hunk 1 above)
Comment (plbossart) [reply src/audio/mixer.c:259]: @abonislawski ping?

**Comment by lgirdwood** [comment]
@abonislawski ping ?

**Comment by lgirdwood** [comment]
@abonislawski ping - need this fix for v1.7

File: src/audio/mixer.c (hunk 1 above)
Comment (abonislawski) [reply src/audio/mixer.c:251]: @plbossart ok updated, sorry for the delay

**Comment by abonislawski** [comment]
@lgirdwood PR updated

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@abonislawski I think we are at limit of byt memory - and checking I can see we are missing a 
```
CONFIG_COMP_TONE=n
```
in it's kconfig (as CHT has this). Can you inlcude this fix in your PR. I think we are good to then merge. Thanks !

**Comment by abonislawski** [comment]
@lgirdwood sure, added this to byt config

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**